rpcs3/Emu/RSX/VK/VKTextureCache.cpp

```diff
@@ -358,4 +374,875 @@ namespace vk
+			linear_format_supported = m_formats_support.bgra8_linear;
+			break;
+		case VK_FORMAT_R8G8B8A8_UNORM:
+			linear_format_supported = m_formats_support.argb8_linear;
+			break;
+		default:
+			rsx_log.error("Unsupported VkFormat 0x%x", static_cast<u32>(format));
+			return nullptr;
+		}
+
+		if (!linear_format_supported)
+		{
+			return nullptr;
+		}
+
+		// Uploads a linear memory range as a BGRA8 texture
+		auto image = std::make_unique<vk::viewable_image>(*m_device, m_memory_types.host_visible_coherent,
+			VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT,
+			VK_IMAGE_TYPE_2D,
+			format,
+			width, height, 1, 1, 1, VK_SAMPLE_COUNT_1_BIT, VK_IMAGE_LAYOUT_PREINITIALIZED,
+			VK_IMAGE_TILING_LINEAR, VK_IMAGE_USAGE_TRANSFER_SRC_BIT | VK_IMAGE_USAGE_SAMPLED_BIT, 0,
+			VMM_ALLOCATION_POOL_UNDEFINED);
+
+		VkImageSubresource subresource{};
+		subresource.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+
+		VkSubresourceLayout layout{};
+		vkGetImageSubresourceLayout(*m_device, image->value, &subresource, &layout);
+
+		void* mem = image->memory->map(0, layout.rowPitch * height);
+
+		auto src = vm::_ptr<const char>(address);
+		auto dst = static_cast<char*>(mem);
+
+		//TODO: SSE optimization
+		for (u32 row = 0; row < height; ++row)
+		{
+			auto casted_src = reinterpret_cast<const be_t<u32>*>(src);
+			auto casted_dst = reinterpret_cast<u32*>(dst);
+
+			for (u32 col = 0; col < width; ++col)
+				casted_dst[col] = casted_src[col];
+
+			src += pitch;
+			dst += layout.rowPitch;
+		}
+
+		image->memory->unmap();
+
+		vk::change_image_layout(cmd, image.get(), VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL);
+
+		auto result = image.get();
+		const u32 resource_memory = width * height * 4; //Rough approximate
+		m_temporary_storage.emplace_back(image);
+		m_temporary_storage.back().block_size = resource_memory;
+		m_temporary_memory_size += resource_memory;
+
+		return result;
+	}
+
+	bool texture_cache::blit(rsx::blit_src_info& src, rsx::blit_dst_info& dst, bool interpolate, vk::surface_cache& m_rtts, vk::command_buffer& cmd)
+	{
+		blitter helper;
+		auto reply = upload_scaled_image(src, dst, interpolate, cmd, m_rtts, helper);
+
+		if (reply.succeeded)
+		{
+			if (reply.real_dst_size)
+			{
+				flush_if_cache_miss_likely(cmd, reply.to_address_range());
+			}
+
+			return true;
+		}
+
+		return false;
+	}
+
+	u32 texture_cache::get_unreleased_textures_count() const
+	{
+		return baseclass::get_unreleased_textures_count() + ::size32(m_temporary_storage);
+	}
+
+	u32 texture_cache::get_temporary_memory_in_use() const
+	{
+		return m_temporary_memory_size;
+	}
+
+	bool texture_cache::is_overallocated() const
+	{
+		const auto total_device_memory = m_device->get_memory_mapping().device_local_total_bytes / 0x100000;
+		u64 quota = 0;
+
+		if (total_device_memory >= 1024)
+		{
+			quota = std::min(3072ull, (total_device_memory * 40) / 100);
+		}
+		else if (total_device_memory >= 768)
+		{
+			quota = 256;
+		}
+		else
+		{
+			quota = std::min(128ull, total_device_memory / 2);
```

Comment:
these std::mins need their type fixed for linux. either with ul instead of ull or with a template arg <u64>